KAFKA-8347: Choose next record to process by timestamp

ableegoldman opened the pull request:
When choosing the next record to process, we should look at the head record's timestamp of each partition and choose the lowest rather than choosing the lowest of the partition's streamtime.

This change effectively makes RecordQueue return the timestamp of the head record rather than its streamtime. Streamtime is removed (replaced) from RecordQueue as it was only being tracked in order to choose the next partition to poll from.

ableegoldman committed:
Use head timestamp instead of partition time

mjsax reviewed (approved):
LGTM -- not sure why Jenkins failed.

Retest this please.

mjsax commented:
Retest this please.

ableegoldman committed:
Added unit test

ableegoldman commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:
```diff
@@ -117,7 +117,7 @@ public void testTimeTracking() {
         // 1:[3, 5]
         // 2:[2, 4, 6]
-        // st: 2
+        // st: 1
```
I noticed the comments throughout this test were not accurate, so I fixed them to reflect the current streamtime logic

ableegoldman committed:
Fixed up broken test, refactored RecordQueue somewhat

ableegoldman commented:
I refactored this somewhat to return RecordQueue.UNKNOWN from timestamp() when the queue is empty, as timestamp() should now be reflecting not stream time but timestamp of the next (head) record. Also fixed some out of date/inaccurate comments in PartitionGroupTest

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:
```diff
@@ -190,24 +190,72 @@ public void testTimeTracking() {
     }
 
+    @Test
+    public void testChooseNextRecord() {
```
nit: should we merge this into existing test? If not, rename to `shouldChooseNextRecordBasedOnHeadTimestampe`

ableegoldman replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:
```diff
@@ -190,24 +190,72 @@ public void testTimeTracking() {
+        // 2:[4, 6]
+        assertEquals(record.timestamp, 2L);
+
+        // get one record, next up should have ts=3 from partition 1 (even though it has larger partition stream time st=5)
```
nit: there is not "partition stream time" -- there is "stream time" and "partition time" :)

ableegoldman replied:
Ack, I chose to use the term to differentiate between the "old" partition time -- largest timestamp observed for a partition so far -- and what is effectively becoming the "new" partition time in this PR, ie timestamp of head record. Will clarify the comment though

ableegoldman committed:
Updated comment and test name

ableegoldman commented:
Closing this in favor of [minor version bump](https://github.com/apache/kafka/pull/6743)

mjsax reviewed (approved):
LGTM.

Call for second review @guozhangwang @bbejeck @vvcephei @cadonna @abbccdda

bbejeck reviewed (approved):
LGTM

bbejeck commented:
Merged #6719 into trunk.

bbejeck commented:
cherry-picked to 2.2 and 2.1
